### ISSUE TITLE
"requireParamDescription" when set to false, it should not flag parameters without trailing hyphens

### ISSUE THREAD

**Issue by tommai78101** [eslint/eslint#10342]
**What rule do you want to change?**

I would like to change the a rule for "valid-jsdoc", `requireParamDescription`.

**Does this change cause the rule to produce more or fewer warnings?**

It should reduce warnings.

**How will the change be implemented? (New option, new default behavior, etc.)?**

The change is to ignore trailing hyphens at the beginning of the parameter description in a JSDoc comment.

This is the intended JSDoc behavior, as noted in this link:

http://usejsdoc.org/tags-param.html

Where JSDoc allows only the name and type for a `@param` entry, like so:

```javascript
/**
 * @param {string} somebody
 */
function sayHello(somebody) {
    alert('Hello ' + somebody);
}
```

**Please provide some example code that this change will affect:**

Using the following code:

```javascript

/**
 * Demo
 * @param {number} a -
 * @param {number} b
 * @param {number} c --------------------------------------------------------
 * @returns {number} Sum
 */
function hello(a, b, c){
	return a + b + c;
}

hello(1, 2, 3);
```
Here in this JSDoc, the example shows parameter 1 is correct, parameter 2 is incorrect, parameter 3 is correct, and the return is correct. Reason why parameter 2 is incorrect, is because it does not have a trailing hyphen. 

The new change should say the description without any trailing hyphens is valid, because there really isn't any requirements for a description. A trailing hyphen is somehow weird, and does not fit the rule's intended purpose of not requiring parameter descriptions, especially with the trailing hyphen.

**What does the rule currently do for this code?**

The current rule states that the parameter description must be required to **at least** insert or append a single trailing hyphen. 

![image](https://user-images.githubusercontent.com/6444116/39941260-fe36d2c6-5529-11e8-9e52-fb119b2e96ff.png)

**What will the rule do after it's changed?**

The new changed rule will allow an empty description to be a valid JSDoc.

**Comment by ilyavolodin** [comment]
If I'm understanding it correctly, you want the following configuration `'valid-jsdoc':['error', { 'requireParamDescription': false }]` to not flag any of the param descriptions in your example, right? I just tried it in the demo, and it seems to be working correctly: [demo](https://eslint.org/demo/#eyJ0ZXh0IjoiLyogZXNsaW50ICd2YWxpZC1qc2RvYyc6WydlcnJvcicsIHsgJ3JlcXVpcmVQYXJhbURlc2NyaXB0aW9uJzogZmFsc2UgfV0gKi9cclxuLyoqXHJcbiAqIERlbW9cclxuICogQHBhcmFtIHtudW1iZXJ9IGEgLVxyXG4gKiBAcGFyYW0ge251bWJlcn0gYlxyXG4gKiBAcGFyYW0ge251bWJlcn0gYyAtLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLVxyXG4gKiBAcmV0dXJucyB7bnVtYmVyfSBTdW1cclxuICovXHJcbmZ1bmN0aW9uIGhlbGxvKGEsIGIsIGMpe1xyXG5cdHJldHVybiBhICsgYiArIGM7XHJcbn1cclxuXHJcbmhlbGxvKDEsIDIsIDMpOyIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6OCwic291cmNlVHlwZSI6Im1vZHVsZSIsImVjbWFGZWF0dXJlcyI6eyJqc3giOnRydWUsImV4cGVyaW1lbnRhbE9iamVjdFJlc3RTcHJlYWQiOnRydWV9fSwicnVsZXMiOnt9LCJlbnYiOnt9fX0=)
I'm I understanding request wrong?

**Comment by tommai78101** [comment]
@ilyavolodin Huh? How come I'm seeing the error in the demo that I took a screenshot of on my end? And it was an online demo. 

Basically, assuming my demo is correct, I want the `'valid-jsdoc':['error', { 'requireParamDescription': false }]` to not flag an error that states the JSDoc parameter description requires at least 1 character in the line.

But now I'm totally confused as to why I have the error, but you don't.      

This is my Visual Studio Code user settings:

```json
{
	"beautify.config": "C:\\Workspace\\.vscode\\.jsbeautifyrc",
	"bracketPairColorizer.forceUniqueOpeningColor": true,
	"bracketPairColorizer.timeOut": 0,
	"diffEditor.renderSideBySide": false,
	"editor.cursorStyle": "line-thin",
	"editor.detectIndentation": false,
	"editor.fontFamily": "Consolas, 'DejaVu Sans Mono', 'MingLiU', 'Meiryo', monospace",
	"editor.fontSize": 11,
	"editor.insertSpaces": false,
	"editor.renderWhitespace": "all",
	"editor.tabSize": 3,
	"eslint.options": {
		"rules": {
			"requireParamDescription": 0,
			"requireReturnDescription": 0,
			"no-extra-parens": 0,
		}
	},
	"html.format.indentInnerHtml": true,
	"typescript.experimental.syntaxFolding": true,
	"window.restoreWindows": "all",
	"workbench.colorCustomizations": {
		"editor.selectionBackground": "#14a",
		"editor.selectionHighlightBackground": "#733",
		"editor.wordHighlightBackground": "#660",
		"editorCursor.foreground": "#fff"
	},
	"workbench.startupEditor": "newUntitledFile",
}
```

I'm wondering if it's because I'm not correctly specifying the rules right, or I'm shooting myself in the foot?

**Comment by tommai78101** [comment]
I mucked around the user settings for LONG while, and I finally understood how to specify the rules correctly.

```json
{
	"beautify.config": "C:\\Workspace\\.vscode\\.jsbeautifyrc",
	"bracketPairColorizer.forceUniqueOpeningColor": true,
	"bracketPairColorizer.timeOut": 0,
	"diffEditor.renderSideBySide": false,
	"editor.cursorStyle": "line-thin",
	"editor.detectIndentation": false,
	"editor.fontFamily": "Consolas, 'DejaVu Sans Mono', 'MingLiU', 'Meiryo', monospace",
	"editor.fontSize": 11,
	"editor.insertSpaces": false,
	"editor.renderWhitespace": "all",
	"editor.tabSize": 3,
	"eslint.options": {
		"rules": {
			"valid-jsdoc": [
				"error",
				{
					"requireParamDescription": false,
					"requireReturnDescription": false,
					"requireReturn": false
				}
			],
			"no-extra-parens": 0
		}
	},
	"html.format.indentInnerHtml": true,
	"typescript.experimental.syntaxFolding": true,
	"window.restoreWindows": "all",
	"workbench.colorCustomizations": {
		"editor.selectionBackground": "#0b4fd8d0",
		"editor.selectionHighlightBackground": "#d1454544",
		"editor.wordHighlightBackground": "#ffff004d",
		"editorCursor.foreground": "#fff"
	},
	"workbench.startupEditor": "newUntitledFile",
}
```

So I learned `rules` is an object that takes key-value pairs. The key refers to rule name, the value is an **array**, not a true/false flag. The first element of the array refers to the warning/error level, and the second element of the array refers to an object that holds the sub-rules of the specified key in `rules`.

This is not something that is clear to me when beginning to use ESLint, but heh, I managed...

Thanks and closing this issue, because in the end, it's a user error.

**Comment by platinumazure** [comment]
Hi @tommai78101, thanks for following up.

I'm sorry that the rule configuration structure wasn't obvious-- I know it's something we can definitely improve. (And I also apologize for not noticing the issue in your last post-- would have been good for us to offer a suggestion there and we failed to respond in a timely manner.)

If you have any suggestions for how we could improve the documentation for first-time ESLint users, please do let us know via a new issue or feel free to submit a pull request. We definitely want to try to make the user experience as good as it can be, and I know we definitely have a ways to go.

Thanks for continuing to use ESLint!

**Comment by tommai78101** [comment]
@platinumazure 

Hi, thanks for the response. I wrote a wiki article in an obscure Microsoft repository for Visual Studio Code here:

https://github.com/Microsoft/vscode-eslint/wiki/Visual-Studio-Code:--Setting-up-ESLint-Rules-(Beginner's-Introduction)

The only suggestion (split into 2 parts) that I have right now is, because setting up ESLint on many different editors all have their own quirks and methods of applying ESLint rules, there really isn't a standardized method of specifying or applying ESLint rules to the way the user wanted.

The hard part is:

It would be nice for ESLint to come up with a standardized rules layout, much akin to C++ standard, ECMAScript standard, or something where it's strictly enforcing a certain syntax format, so all editors must apply the same strict syntax when integrating ESLint into their workflows.

The easier part is:

The documentation should have an article of a general list of supported syntaxes for writing ESLint rules. Some editors support configuration files written in JSON for ESLint, so JSON syntaxes should be one of the many supported syntaxes on that article. There are some that are Batch files based, Bash/Shell scripts, .editorconfig based, or even INI files.

Again, it's the breath of configuration syntaxes that are too broad for me to search for the right syntaxes suitable for the editor I'm using. That's the issue right now for me. But maybe one of the contributors know more than me and knows how to address this issue going forward.

That's all.

**Comment by platinumazure** [comment]
Thanks for taking the time to write a detailed response, @tommai78101. I'll think on it and see what we can do.
